lib/web_ui/lib/src/engine/text/layout_service.dart

```diff
@@ -218,8 +224,24 @@ class TextLayoutService {
       if (longestLineWidth < line.width) {
         longestLine = line;
       }
+
+      final double left = line.left;
+      if (left < boundsLeft) {
+        boundsLeft = left;
+      }
+      final double right = left + line.width;
+      if (right > boundsRight) {
+        boundsRight = right;
+      }
     }
 
+    _paintBounds = ui.Rect.fromLTRB(
+      boundsLeft,
+      0,
```

Comment:
Good point!

The previous version already used the `(0, paragraph.height)` range for the vertical bounds (see `recording_canvas.dart`, where `offset.dy` was used as the `top` and `top + engineParagraph.height` as the `bottom`). That's still the case after this PR.

What is changing are the horizontal bounds. Your concern is still valid for the horizontal case (what if a glyph paints outside its layout bounds due to some [swash](https://en.wikipedia.org/wiki/Swash_(typography))). Currently `dart:ui` doesn't provide this information to the framework, so this problem already exists (e.g. the framework may clip or prune contents, thinking that it's not visible). This can probably be worked around by the developer by using extra padding if they know that they are using some font that results in this kind of out-of-bounds painting, and I think horizontal extrusions are less common than vertical ones due to the risk of overlapping with adjacent glyphs.

Maybe this is largely solvable by adding some padding proactively? For example, we could add a certain % of the font size on both sides. WDYT? @mdebbar I'm curious about your thoughts on the topic too.

Reply:
I agree with Yegor. Weird characters that paint outside of their bounds are already broken because of clipping, etc.

At this time, I would say it's fair to assume that the top paint bound is 0, especially that there's no concept of vertical alignment (yet) in paragraphs.

Reply:
I don't know enough about this, so 🚀